### PR TITLE
Simplify the user delete service and it's tests

### PR DESCRIPTION
**Pull request by jon-betts**
Requires:

 * https://github.com/hypothesis/h/pull/8075
 * https://github.com/hypothesis/h/pull/8082

Refactor the user service to be smaller and normaller. This:

 * Adds test for the factory method
 * Keeps a DB session instead of the whole request object
 * Simplifies the code a lot
 * Simplifies the tests a bit
 * Moves to SQLAlchemy 2.0 friendly style

## Review notes

There should be no change in behavior, apart from being slightly faster.

The main change here is to replace a lot of Python code with SQL. Previously we were getting the list of groups created by the user, and then searching for those groups which have annotations in them by other users. After this we did a lot of list comprehensions and set manipulation to try and separate out two sets of groups:

 * Those which the user created, with no annotations by other people - These get deleted
 * Those which the user created, with annotations by other people - These get the creator set to `None`

In this version we look for the groups the user created and simultaneously count all the annotations by other users with an outer join all in SQL. We then take appropriate action depending on whether the count is 0 or not.

## If this works, why bother?

We are going to be letting the user trigger this code in the future ideally. I wanted to read this and understand what it did to make sure it was ready for to be used in this way.

I found the previous formulation long and confusing.

## Testing notes

This doesn't cover all the nitty gritty with different groups and memberships. Hopefully the unit tests do that, but we can kick the tires to make sure it's basically sound:

 * `make services devdata dev`
 * Visit: http://localhost:5000/login
 * Login as `devdata_admin` / `pass`
 * Visit: http://localhost:5000/admin/users
 * Search for `devdata_user`
 * Press "Delete user" and confirm
 * Nothing explodes!
 * Search for `devdata_user`. They should be gone
 * `make devdata` to reset

### CODE REVIEW

File: h/services/user_delete.py [hunk 1]
```diff
@@ -15,61 +22,33 @@ def delete(self, user):
...
-            return []
-
-        query = (
-            self.request.db.query(Annotation.groupid)
-            .filter(Annotation.groupid.in_(group_ids), Annotation.userid != user.userid)
-            .group_by(Annotation.groupid)
-        )
-        groupids_with_other_user_anns = [pubid for (pubid,) in query.all()]
-
-        return [g for g in groups if g.pubid in groupids_with_other_user_anns]
```
Comment (marcospri) [inline h/services/user_delete.py:56]: Oh this is hard to follow but I reckon you found the right SQL alternative to do it.

**Comment by marcospri** [comment]
The PR is not changing this but wee seem to have this fk to `user`

```
SELECT conrelid::regclass AS table_name, 
       conname AS foreign_key, 
       pg_get_constraintdef(oid) 
FROM   pg_constraint 
WHERE  contype = 'f' and pg_get_constraintdef(oid) like 'FOREIGN KEY (user_id) REFERENCES "user"(id)%'                                                                      
AND    connamespace = 'public'::regnamespace
ORDER  BY conrelid::regclass::text, contype DESC;
     table_name     |              foreign_key              |                     pg_get_constraintdef                      
--------------------+---------------------------------------+---------------------------------------------------------------
 authticket         | fk__authticket__user_id__user         | FOREIGN KEY (user_id) REFERENCES "user"(id) ON DELETE CASCADE
 authzcode          | fk__authzcode__user_id__user          | FOREIGN KEY (user_id) REFERENCES "user"(id) ON DELETE CASCADE
 featurecohort_user | fk__featurecohort_user__user_id__user | FOREIGN KEY (user_id) REFERENCES "user"(id)
 flag               | fk__flag__user_id__user               | FOREIGN KEY (user_id) REFERENCES "user"(id) ON DELETE CASCADE
 user_group         | fk__user_group__user_id__user         | FOREIGN KEY (user_id) REFERENCES "user"(id)
 user_identity      | fk__user_identity__user_id__user      | FOREIGN KEY (user_id) REFERENCES "user"(id) ON DELETE CASCADE
(6 rows)
```

I don't know H' data model enough but I guess we are fine to rely on the cascade on this.

**Review by marcospri — APPROVED** [review]

**Comment by jon-betts** [comment]
> I don't know H' data model enough but I guess we are fine to rely on the cascade on these.

Yes, I couldn't see anything explicitly deleting "group_membership" for example, but when I tested it, they do go.
